src/commands/importFromRegistry.ts

```diff
@@ -0,0 +1,168 @@
+
+  const config = vscode.workspace.getConfiguration("intersystems", scope);
+  const serverDefinitions: any = config.get("servers");
+
+  const newServerNames = new Array<string>();
+  const serversMissingUsernames = new Array<string>();
+  const cancellationToken = new vscode.CancellationTokenSource().token;
+  cancellationToken.onCancellationRequested(() => {
+    vscode.window.showInformationMessage("Cancelled server import.");
+  });
+
+  await cmd.get(
+    "reg query " +
+      hkeyLocalMachine +
+      "\\SOFTWARE\\Intersystems\\Cache\\Servers",
+    async (err, data, stderr) => {
+      // Prompt user for username (if not in registry) and password
+      data.split("\n").forEach((serverName) => {
+        // remove HKEY_LOCAL_MACHINE\ and whitespace from the server name
+        const path = serverName.substring(hkeyLocalMachine.length + 1).trim();
+
+        const originalName: string = serverName.split("\\").pop().trim();
+        // Enforce the rules from package.json on the server name
+        const name = originalName.toLowerCase().replace(/[^a-z0-9-._~]/g, "~");
+        const getProperty = (property: string) =>
+          vsWinReg.GetStringRegKey(hkeyLocalMachine, path, property);
+
+        if (name !== "" && !config.has("servers." + name)) {
+          newServerNames.push(name);
+          let username: string | undefined;
+          try {
+            username = vsWinReg.GetStringRegKey(
+              "HKEY_CURRENT_USER",
+              "Software\\InterSystems\\Cache\\Servers\\" + originalName,
+              "Server User Name",
+            );
+          } catch (e) {
+            // No-op; will assume the key did not exist (it is valid for it not to).
+          }
+
+          if (username === undefined || username === "") {
+            serversMissingUsernames.push(name);
+          }
+
+          const usesHttps = getProperty("HTTPS") === "1";
+          serverDefinitions[name] = {
+            description: getProperty("Comment"),
+            username: (username === undefined || username === "" ? undefined : username),
+            webServer: {
+              host: getProperty("WebServerAddress") || getProperty("Address"),
+              pathPrefix: getProperty("WebServerInstanceName"),
+              port: parseInt(getProperty("WebServerPort"), 10),
+              scheme: usesHttps ? "https" : "http",
+            },
+          };
+        }
+      });
+
+      if (serversMissingUsernames.length) {
+        let serverName = serversMissingUsernames[0];
+        let username = await vscode.window.showInputBox({
+          ignoreFocusOut: true,
+          placeHolder: "username",
+          prompt: `Username for server '${serverName}'`,
+          validateInput: ((value) => {
+              return value.length > 0 ? "" : "Mandatory field";
+          }),
+        }, cancellationToken);
+        if (cancellationToken.isCancellationRequested || username === undefined) {
+          return;
+        }
```

Comment:
At this point I think we need `serverDefinitions[serverName].username = username`

Otherwise in the 1-server-to-import case this never gets stored in settings.json, and also causes the text 'undefined' to appear in a subsequent quickpick (see later comment).